File: prince-of-versions/src/test/java/co/infinum/princeofversions/PresenterTest.java
package co.infinum.princeofversions;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentMatchers;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executor;

import co.infinum.princeofversions.mocks.SingleThreadExecutor;

import static org.assertj.core.api.Java6Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyBoolean;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class PresenterTest {

    private static final Map<String, String> DEFAULT_METADATA = new HashMap<>();

    @Mock
    Interactor interactor;

    @Mock
    Storage storage;

    @Mock
    Loader loader;

    @Mock
    ApplicationConfiguration appConfig;

    @Mock
    UpdateInfo updateInfo;

    private PresenterImpl presenter;

    @Before
    public void setUp() {
        presenter = new PresenterImpl(interactor, storage);
    }

    @After
    public void tearDown() {
        presenter = null;
    }

    @Test
    public void testMandatoryUpdate() throws Throwable {
        CheckResult checkResult = CheckResult.mandatoryUpdate(10, DEFAULT_METADATA, updateInfo);
        when(interactor.check(any(Loader.class), any(ApplicationConfiguration.class))).thenReturn(checkResult);

        UpdateResult result = presenter.run(loader, appConfig);

        assertThat(result).isEqualTo(new UpdateResult(
            updateInfo,
            DEFAULT_METADATA,
            UpdateStatus.REQUIRED_UPDATE_NEEDED,
            10
        ));

        verify(storage, times(1)).rememberLastNotifiedVersion(checkResult.getUpdateVersion());
    }

    @Test
    public void testNoUpdate() throws Throwable {
        CheckResult checkResult = CheckResult.noUpdate(10, DEFAULT_METADATA, updateInfo);
        when(interactor.check(any(Loader.class), any(ApplicationConfiguration.class))).thenReturn(checkResult);

        UpdateResult result = presenter.run(loader, appConfig);

        assertThat(result).isEqualTo(new UpdateResult(
            updateInfo,
            DEFAULT_METADATA,
            UpdateStatus.NO_UPDATE_AVAILABLE,
            checkResult.getUpdateVersion()
        ));

        verify(storage, times(0)).rememberLastNotifiedVersion(anyInt());
    }

    @Test
    public void testOptionalUpdateWhenNotNotifiedUpdate() throws Throwable {
        CheckResult checkResult = CheckResult.optionalUpdate(12, NotificationType.ONCE, DEFAULT_METADATA, updateInfo);
        when(storage.lastNotifiedVersion(null)).thenReturn(11);
        when(interactor.check(any(Loader.class), any(ApplicationConfiguration.class))).thenReturn(checkResult);

        UpdateResult result = presenter.run(loader, appConfig);

        assertThat(result).isEqualTo(new UpdateResult(
            updateInfo,
            DEFAULT_METADATA,
            UpdateStatus.NEW_UPDATE_AVAILABLE,
            checkResult.getUpdateVersion()
        ));

        verify(storage, times(1)).rememberLastNotifiedVersion(checkResult.getUpdateVersion());
    }

    @Test
    public void testMandatoryUpdateAlreadyNotifiedUpdateWithAlways() throws Throwable {
        CheckResult checkResult = CheckResult.optionalUpdate(12, NotificationType.ALWAYS, DEFAULT_METADATA, updateInfo);
        when(storage.lastNotifiedVersion(null)).thenReturn(12);
        when(interactor.check(any(Loader.class), any(ApplicationConfiguration.class))).thenReturn(checkResult);

        UpdateResult result = presenter.run(loader, appConfig);

        assertThat(result).isEqualTo(new UpdateResult(
            updateInfo,
            DEFAULT_METADATA,
            UpdateStatus.NEW_UPDATE_AVAILABLE,
            checkResult.getUpdateVersion()
        ));

        verify(storage, times(1)).rememberLastNotifiedVersion(checkResult.getUpdateVersion());
    }

    @Test
    public void testMandatoryUpdateAlreadyNotifiedUpdateWithOnce() throws Throwable {
        CheckResult checkResult = CheckResult.optionalUpdate(12, NotificationType.ONCE, DEFAULT_METADATA, updateInfo);
        when(storage.lastNotifiedVersion(null)).thenReturn(12);
        when(interactor.check(any(Loader.class), any(ApplicationConfiguration.class))).thenReturn(checkResult);

        UpdateResult result = presenter.run(loader, appConfig);

        assertThat(result).isEqualTo(new UpdateResult(
            updateInfo,
            DEFAULT_METADATA,
            UpdateStatus.NO_UPDATE_AVAILABLE,
            checkResult.getUpdateVersion()
        ));

        verify(storage, times(0)).rememberLastNotifiedVersion(anyInt());
    }

    @Test
    public void testSyncCheckMandatory() throws Throwable {
        PresenterImpl mock = mock(PresenterImpl.class);
        UpdateResult expected = new UpdateResult(updateInfo, DEFAULT_METADATA, UpdateStatus.REQUIRED_UPDATE_NEEDED, 10);
        when(mock.run(any(Loader.class), any(ApplicationConfiguration.class))).thenReturn(expected);
        when(mock.check(any(Loader.class), any(ApplicationConfiguration.class))).thenCallRealMethod();

        UpdateResult actual = mock.check(loader, appConfig);

        assertThat(actual).isEqualTo(expected);
    }

    @Test
    public void testSyncCheckOptional() throws Throwable {
        PresenterImpl mock = mock(PresenterImpl.class);
        UpdateResult expected = new UpdateResult(updateInfo, DEFAULT_METADATA, UpdateStatus.NEW_UPDATE_AVAILABLE, 10);
        when(mock.run(any(Loader.class), any(ApplicationConfiguration.class))).thenReturn(expected);
        when(mock.check(any(Loader.class), any(ApplicationConfiguration.class))).thenCallRealMethod();

        UpdateResult actual = mock.check(loader, appConfig);

        assertThat(actual).isEqualTo(expected);
    }

    @Test
    public void testSyncCheckNoUpdate() throws Throwable {
        PresenterImpl mock = mock(PresenterImpl.class);
        UpdateResult expected = new UpdateResult(updateInfo, DEFAULT_METADATA, UpdateStatus.NO_UPDATE_AVAILABLE, 10);
        when(mock.run(any(Loader.class), any(ApplicationConfiguration.class))).thenReturn(expected);
        when(mock.check(any(Loader.class), any(ApplicationConfiguration.class))).thenCallRealMethod();

        UpdateResult actual = mock.check(loader, appConfig);

        assertThat(actual).isEqualTo(expected);
    }

    @Test
    public void testAsyncCheckMandatory() throws Throwable {
        UpdaterCallback callback = mock(UpdaterCallback.class);
        PresenterImpl mock = mock(PresenterImpl.class);
        UpdateResult expected =
            new UpdateResult(updateInfo, DEFAULT_METADATA, UpdateStatus.REQUIRED_UPDATE_NEEDED, 10);
        when(mock.run(any(Loader.class), any(ApplicationConfiguration.class))).thenReturn(expected);
        when(mock.check(any(Loader.class), any(Executor.class), any(UpdaterCallback.class), any(ApplicationConfiguration.class)))
            .thenCallRealMethod();
        when(mock.createCall()).thenCallRealMethod();

        mock.check(loader, new SingleThreadExecutor(), callback, appConfig);

        verify(callback, times(1)).onSuccess(expected);
        verify(callback, times(0)).onError(any(Throwable.class));
    }

    @Test
    public void testAsyncCheckOptional() throws Throwable {
        UpdaterCallback callback = mock(UpdaterCallback.class);
        PresenterImpl mock = mock(PresenterImpl.class);
        UpdateResult expected = new UpdateResult(updateInfo, DEFAULT_METADATA, UpdateStatus.NEW_UPDATE_AVAILABLE, 10);
        when(mock.run(any(Loader.class), any(ApplicationConfiguration.class))).thenReturn(expected);
        when(mock.check(any(Loader.class), any(Executor.class), any(UpdaterCallback.class), any(ApplicationConfiguration.class)))
            .thenCallRealMethod();
        when(mock.createCall()).thenCallRealMethod();

        mock.check(loader, new SingleThreadExecutor(), callback, appConfig);

        verify(callback, times(1)).onSuccess(expected);
        verify(callback, times(0)).onError(any(Throwable.class));
    }

    @Test
    public void testAsyncCheckNoUpdate() throws Throwable {
        UpdaterCallback callback = mock(UpdaterCallback.class);
        PresenterImpl mock = mock(PresenterImpl.class);
        UpdateResult expected = new UpdateResult(updateInfo, DEFAULT_METADATA, UpdateStatus.NO_UPDATE_AVAILABLE, 10);
        when(mock.run(any(Loader.class), any(ApplicationConfiguration.class))).thenReturn(expected);
        when(mock.check(any(Loader.class), any(Executor.class), any(UpdaterCallback.class), any(ApplicationConfiguration.class)))
            .thenCallRealMethod();
        when(mock.createCall()).thenCallRealMethod();

        mock.check(loader, new SingleThreadExecutor(), callback, appConfig);

        verify(callback, times(1)).onSuccess(expected);
        verify(callback, times(0)).onError(any(Throwable.class));
    }

    @Test
    public void testAsyncCheckNoUpdateCancel() throws Throwable {
        UpdaterCallback callback = mock(UpdaterCallback.class);
        PresenterImpl mock = mock(PresenterImpl.class);
        UpdateResult expected = new UpdateResult(updateInfo, DEFAULT_METADATA, UpdateStatus.NO_UPDATE_AVAILABLE, 10);
        when(mock.run(any(Loader.class), any(ApplicationConfiguration.class))).thenReturn(expected);
        when(mock.check(any(Loader.class), any(Executor.class), any(UpdaterCallback.class), any(ApplicationConfiguration.class)))
            .thenCallRealMethod();
        when(mock.createCall()).thenReturn(new UpdaterCancelable() {
            @Override
            public boolean isCanceled() {
                return true;
            }
        });

        mock.check(loader, new SingleThreadExecutor(), callback, appConfig);

        verify(callback, times(0)).onSuccess(expected);
        verify(callback, times(0)).onError(any(Throwable.class));
    }

    @Test
    public void testAsyncCheckMandatoryUpdateCancel() throws Throwable {
        UpdaterCallback callback = mock(UpdaterCallback.class);
        PresenterImpl mock = mock(PresenterImpl.class);
        UpdateResult expected = new UpdateResult(updateInfo, DEFAULT_METADATA, UpdateStatus.REQUIRED_UPDATE_NEEDED, 10);
        when(mock.run(any(Loader.class), any(ApplicationConfiguration.class))).thenReturn(expected);
        when(mock.check(any(Loader.class), any(Executor.class), any(UpdaterCallback.class), any(ApplicationConfiguration.class)))
            .thenCallRealMethod();
        when(mock.createCall()).thenReturn(new UpdaterCancelable() {
            @Override
            public boolean isCanceled() {
                return true;
            }
        });

        mock.check(loader, new SingleThreadExecutor(), callback, appConfig);

        verify(callback, times(0)).onSuccess(expected);
        verify(callback, times(0)).onError(any(Throwable.class));
    }

    @Test
    public void testAsyncCheckOptionalUpdateCancel() throws Throwable {
        UpdaterCallback callback = mock(UpdaterCallback.class);
        PresenterImpl mock = mock(PresenterImpl.class);
        UpdateResult expected = new UpdateResult(updateInfo, DEFAULT_METADATA, UpdateStatus.NEW_UPDATE_AVAILABLE, 10);
        when(mock.run(any(Loader.class), any(ApplicationConfiguration.class))).thenReturn(expected);
        when(mock.check(any(Loader.class), any(Executor.class), any(UpdaterCallback.class), any(ApplicationConfiguration.class)))
            .thenCallRealMethod();
        when(mock.createCall()).thenReturn(new UpdaterCancelable() {
            @Override
            public boolean isCanceled() {
                return true;
            }
        });

        mock.check(loader, new SingleThreadExecutor(), callback, appConfig);

        verify(callback, times(0)).onSuccess(expected);
        verify(callback, times(0)).onError(any(Throwable.class));
    }

    @Test
    public void testAsyncCheckError() throws Throwable {
        UpdaterCallback callback = mock(UpdaterCallback.class);
        PresenterImpl mock = mock(PresenterImpl.class);
        Throwable throwable = new IllegalStateException();
        when(mock.run(any(Loader.class), any(ApplicationConfiguration.class))).thenThrow(throwable);
        when(mock.check(any(Loader.class), any(Executor.class), any(UpdaterCallback.class), any(ApplicationConfiguration.class)))
            .thenCallRealMethod();
        when(mock.createCall()).thenCallRealMethod();

        mock.check(loader, new SingleThreadExecutor(), callback, appConfig);

        verify(callback, times(0)).onSuccess(any(UpdateResult.class));
        verify(callback, times(1)).onError(any(Throwable.class));
    }

    @Test(expected = IllegalStateException.class)
    public void testSyncCheckError() throws Throwable {
        when(interactor.check(any(Loader.class), any(ApplicationConfiguration.class))).thenThrow(new IllegalStateException());
        presenter.check(loader, appConfig);
    }
}

File: prince-of-versions/src/test/java/co/infinum/princeofversions/PresenterTest.kt
package co.infinum.princeofversions

import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.Mock
import org.mockito.junit.MockitoJUnitRunner
import org.mockito.kotlin.any
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.times
import org.mockito.kotlin.verify
import org.mockito.kotlin.whenever
import java.util.concurrent.Executor

@RunWith(MockitoJUnitRunner::class)
class PresenterTest {

    @Mock
    private lateinit var interactor: Interactor

    @Mock
    private lateinit var storage: Storage

    @Mock
    private lateinit var loader: Loader

    @Mock
    private lateinit var appConfig: ApplicationConfiguration

    @Mock
    private lateinit var updateInfo: UpdateInfo

    private lateinit var presenter: PresenterImpl

    private val defaultMetadata: Map<String, String> = emptyMap()

    @Before
    fun setUp() {
        presenter = PresenterImpl(interactor, storage)
    }

    @Test
    fun testMandatoryUpdate() {
        val checkResult = CheckResult.mandatoryUpdate(10, defaultMetadata, updateInfo)
        whenever(interactor.check(any(), any())).thenReturn(checkResult)

        val result = presenter.run(loader, appConfig)

        val expected = UpdateResult(updateInfo, defaultMetadata, UpdateStatus.REQUIRED_UPDATE_NEEDED, 10)
        assertThat(result).isEqualTo(expected)
        verify(storage, times(1)).rememberLastNotifiedVersion(checkResult.updateVersion)
    }

    @Test
    fun testNoUpdate() {
        val checkResult = CheckResult.noUpdate(10, defaultMetadata, updateInfo)
        whenever(interactor.check(any(), any())).thenReturn(checkResult)

        val result = presenter.run(loader, appConfig)

        val expected = UpdateResult(updateInfo, defaultMetadata, UpdateStatus.NO_UPDATE_AVAILABLE, checkResult.updateVersion)
        assertThat(result).isEqualTo(expected)
        verify(storage, never()).rememberLastNotifiedVersion(anyInt())
    }

    @Test
    fun testOptionalUpdateFirstTime() {
        val checkResult = CheckResult.optionalUpdate(12, NotificationType.ONCE, defaultMetadata, updateInfo)
        whenever(storage.lastNotifiedVersion(null)).thenReturn(null) // First time seeing any update
        whenever(interactor.check(any(), any())).thenReturn(checkResult)

        val result = presenter.run(loader, appConfig)

        val expected = UpdateResult(updateInfo, defaultMetadata, UpdateStatus.NEW_UPDATE_AVAILABLE, checkResult.updateVersion)
        assertThat(result).isEqualTo(expected)
        verify(storage, times(1)).rememberLastNotifiedVersion(checkResult.updateVersion)
    }

    @Test
    fun testOptionalUpdateWhenNotNotified() {
        val checkResult = CheckResult.optionalUpdate(12, NotificationType.ONCE, defaultMetadata, updateInfo)
        whenever(storage.lastNotifiedVersion(null)).thenReturn(11) // Old version notified
        whenever(interactor.check(any(), any())).thenReturn(checkResult)

        val result = presenter.run(loader, appConfig)

        val expected = UpdateResult(updateInfo, defaultMetadata, UpdateStatus.NEW_UPDATE_AVAILABLE, checkResult.updateVersion)
        assertThat(result).isEqualTo(expected)
        verify(storage, times(1)).rememberLastNotifiedVersion(checkResult.updateVersion)
    }

    @Test
    fun testOptionalUpdateNotifiedAlways() {
        val checkResult = CheckResult.optionalUpdate(12, NotificationType.ALWAYS, defaultMetadata, updateInfo)
        whenever(storage.lastNotifiedVersion(null)).thenReturn(12) // Same version notified
        whenever(interactor.check(any(), any())).thenReturn(checkResult)

        val result = presenter.run(loader, appConfig)

        val expected = UpdateResult(updateInfo, defaultMetadata, UpdateStatus.NEW_UPDATE_AVAILABLE, checkResult.updateVersion)
        assertThat(result).isEqualTo(expected)
        verify(storage, times(1)).rememberLastNotifiedVersion(checkResult.updateVersion)
    }

    @Test
    fun testOptionalUpdateNotifiedOnce() {
        val checkResult = CheckResult.optionalUpdate(12, NotificationType.ONCE, defaultMetadata, updateInfo)
        whenever(storage.lastNotifiedVersion(null)).thenReturn(12) // Same version notified
        whenever(interactor.check(any(), any())).thenReturn(checkResult)

        val result = presenter.run(loader, appConfig)

        val expected = UpdateResult(updateInfo, defaultMetadata, UpdateStatus.NO_UPDATE_AVAILABLE, checkResult.updateVersion)
        assertThat(result).isEqualTo(expected)
        verify(storage, never()).rememberLastNotifiedVersion(anyInt())
    }

    @Test
    fun testSyncCheckError() {
        whenever(interactor.check(any(), any())).thenThrow(IllegalStateException())
        assertThatThrownBy { presenter.check(loader, appConfig) }
            .isInstanceOf(IllegalStateException::class.java)
    }

    @Test
    fun testAsyncCheckSuccess() {
        val callback: UpdaterCallback = mock()
        val executor = Executor { it.run() }
        val checkResult = CheckResult.mandatoryUpdate(10, defaultMetadata, updateInfo)
        val expected = UpdateResult(updateInfo, defaultMetadata, UpdateStatus.REQUIRED_UPDATE_NEEDED, 10)
        whenever(interactor.check(loader, appConfig)).thenReturn(checkResult)

        presenter.check(loader, executor, callback, appConfig)

        verify(callback, times(1)).onSuccess(expected)
        verify(callback, never()).onError(any())
    }

    @Test
    fun testAsyncCheckError() {
        val callback: UpdaterCallback = mock()
        val executor = Executor { it.run() }
        val throwable = IllegalStateException()
        whenever(interactor.check(loader, appConfig)).thenThrow(throwable)

        presenter.check(loader, executor, callback, appConfig)

        verify(callback, never()).onSuccess(any())
        verify(callback, times(1)).onError(throwable)
    }

    @Test
    fun testAsyncCheckSuccessIsIgnoredWhenCanceled() {
        val callback: UpdaterCallback = mock()
        val executor = Executor { /* Don't run immediately */ }

        val cancelable = presenter.check(loader, executor, callback, appConfig)
        cancelable.cancel()

        verify(callback, never()).onSuccess(any())
        verify(callback, never()).onError(any())
    }

    @Test
    fun testAsyncCheckErrorIsIgnoredWhenCanceled() {
        val callback: UpdaterCallback = mock()
        val executor = Executor { /* Don't run immediately */ }

        val cancelable = presenter.check(loader, executor, callback, appConfig)
        cancelable.cancel()

        verify(callback, never()).onSuccess(any())
        verify(callback, never()).onError(any())
    }
}
